lang/java/avro/src/main/java/org/apache/avro/Schema.java

```diff
@@ -751,8 +751,32 @@ public void writeName(Names names, JsonGenerator gen) throws IOException {
     }
 
     public String getQualified(String defaultSpace) {
-      return (space == null || space.equals(defaultSpace)) ? name : full;
+      return this.shouldWriteFull(defaultSpace) ? full : name;
     }
+
+    /**
+     * Determine if full name must be written. There are 2 cases for true :
+     * defaultSpace != from this.space or name is already a Schema.Type (int, array
+     * ...)
+     *
+     * @param defaultSpace : default name space.
+     * @return true if full name must be written.
+     */
+    private boolean shouldWriteFull(String defaultSpace) {
+      if (space != null && space.equals(defaultSpace)) {
+        try {
+          // name is a 'Type', so namespace must be written (int should return true, Int
+          // should return false).
+          return Type.valueOf(this.name.toUpperCase(Locale.ENGLISH)).name.equals(this.name);
+        } catch (IllegalArgumentException ex) {
+          // namespace can be omitted, (default & name is not a type)
+          return false;
+        }
```

Comment:
This code catches the exception as flow control, which is quite expensive. I think we should prefer a comparison loop instead:
```
      for (Type type : Type.values()) {
        if (type.name.equals(name)) {
          return true;
        }
      }
      return false;
```

Reply:
Well, Type.valueOf use an internal hashmap (see Class.enumConstantDirectory method), that will have a direct access to Type values ...
And, as reminded [here](https://medium.com/javarevisited/micro-optimizations-in-java-good-nice-and-slow-enum-261e6f77bd2e), Type.values() method will build a clone of all values.
So, the code i proposed will be faster when exist (true return without exception), but i don't know for exception creation vs Type.values() with clone.
Tell me what you think, i can change code.

Reply:
But anyway, for 14 values, i think you're right, i updated the code.